internal/task_documentation/task_documentation.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2022 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+const DirectoryPermissions = 0755
+
+var yamlDir = flag.String("yaml", "", "Location of the generated YAML files")
+var startPath = flag.String("startPath", "", "Location of path to start from")
+var taskDir = flag.String("taskDir", "task", "Location of tekton tasks to scrape")
+
+func main() {
+	flag.Parse()
+	var err error
+	defer func() {
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+	}()
+
+	// YAML
+	if *yamlDir != "" {
+		if err = os.MkdirAll(*yamlDir, DirectoryPermissions); err != nil {
+			return
+		}
+		if err := findTektonTasks(*startPath, *taskDir, *yamlDir); err != nil {
+			return
+		}
+	}
+}
+
+func findTektonTasks(basePath, taskDir, dstDir string) error {
+	var err error
+	if basePath == "" {
+		basePath, err = getGitRoot()
+		if err != nil {
+			return fmt.Errorf("no base path provided and there was an error getting Git root: %v", err)
+		}
+	}
+
+	taskPath := filepath.Join(basePath, taskDir)
+	err = filepath.Walk(taskPath, func(currentPath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() && strings.HasSuffix(currentPath, ".yaml") {
+			isTektonTask, err := isTektonTask(currentPath)
+			if err != nil {
+				return err
+			}
+			if isTektonTask {
+				err = copyFile(currentPath, filepath.Join(basePath, dstDir, info.Name()))
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+	return err
+}
+
+func getGitRoot() (string, error) {
+	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
+	output, err := cmd.Output()
+	if err != nil {
+		return "", err
+	}
+	scanner := bufio.NewScanner(strings.NewReader(string(output)))
+	for scanner.Scan() {
+		return scanner.Text(), nil
+	}
+	return "", nil
+}
+
+func copyFile(src, dst string) error {
+	fmt.Printf("Copying %s to %s\n", src, dst)
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+
+	out, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	_, err = io.Copy(out, in)
+	if err != nil {
+		return err
+	}
+	return out.Close()
```

Comment:
Isn't `out.Close()` handled by the `defer` a few lines above?

Reply:
So the `defer out.Close()` calls are handling errors on open and create. The `return out.Close()` is ensuring that the file is closed after copying.

Reply:
I don't think that's how defer works. `defer out.Close()` will make `out.Close()` be called at the end of the function, regardless of where it returns from.

Reply:
You're correct, the `defer` does handle the closing. My thought with the final `return out.Close()` would be a bit of an explicit error catching. If we rely just on the `defer`, this function will provide the error regarding opening the input file or creating the output file to the caller, but if closing the output file has an error, we end up not passing that back if we don't return the result of calling `out.Close()`

Reply:
I understanding your reasoning, but the [docs](https://pkg.go.dev/os#File.Close) say this:
> Close will return an error if it has already been called. 

The happy path on this function will always call `Close` twice, and will always generate an error on the second call, but that error is ignored. :shrug:

Reply:
Well then, no need for that. :man_shrugging: